### PR TITLE
Add elliptic AnalyticSolution boundary condition

### PR DESCRIPTION
**Pull request by nilsvu**
## Proposed changes

This boundary condition can be used by any elliptic system to impose Dirichlet or Neumann boundary conditions from an analytic solution.

### Code review checklist

- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).
- [ ] The PR lists upgrade instructions and is labeled `bugfix` or
  `major new feature` if appropriate.

### CODE REVIEW

**Review by fmahebert — DISMISSED** [review]

File: src/Elliptic/BoundaryConditions/AnalyticSolution.hpp [hunk 1]
```diff
@@ -0,0 +1,234 @@
...
+  /// \cond
+  explicit AnalyticSolution(CkMigrateMessage* m) noexcept : Base(m) {}
+  using PUP::able::register_constructor;
+  WRAPPED_PUPable_decl_template(AnalyticSolution);
+  /// \endcond
+
+  /// Select which `elliptic::BoundaryConditionType` to apply for each field
+  explicit AnalyticSolution(
+      const typename elliptic::OptionTags::BoundaryConditionType<
+          FieldTags>::type... boundary_condition_types) noexcept
```
Comment (nilsdeppe) [inline src/Elliptic/BoundaryConditions/AnalyticSolution.hpp:103]: I would find this a lot easier to read if you wrote out the type. Does that not work because pack expansion? If that's the case, could you at least document what the type is?

File: src/Elliptic/BoundaryConditions/AnalyticSolution.hpp [hunk 2]
```diff
@@ -0,0 +1,234 @@
...
+                 ::Tags::Normalized<domain::Tags::UnnormalizedFaceNormal<
+                     Dim, Frame::Inertial>>>;
+  using volume_tags =
+      tmpl::list<::Tags::AnalyticSolutionsBase, domain::Tags::Mesh<Dim>>;
+
+  template <typename OptionalAnalyticSolutions>
+  void apply(const OptionalAnalyticSolutions& optional_analytic_solutions,
+             const Mesh<Dim>& volume_mesh, const Direction<Dim>& direction,
+             const tnsr::i<DataVector, Dim>& face_normal,
+             const gsl::not_null<typename FieldTags::type*>... fields,
```
Comment (nilsdeppe) [inline src/Elliptic/BoundaryConditions/AnalyticSolution.hpp]: We put `not_null` arguments at the front of the list

File: src/Elliptic/BoundaryConditions/AnalyticSolution.hpp [hunk 3]
```diff
@@ -0,0 +1,234 @@
...
+             const gsl::not_null<typename FieldTags::type*>... n_dot_fluxes)
+      const noexcept {
+    const auto& analytic_solutions =
+        [&optional_analytic_solutions]() noexcept -> decltype(auto) {
```
Comment (nilsdeppe) [inline src/Elliptic/BoundaryConditions/AnalyticSolution.hpp]: why do you need to specify `decltype(auto)` as the return type?

File: src/Elliptic/BoundaryConditions/AnalyticSolution.hpp [hunk 4]
```diff
@@ -0,0 +1,234 @@
...
+      if constexpr (tt::is_a_v<std::optional, OptionalAnalyticSolutions>) {
+        if (not optional_analytic_solutions.has_value()) {
+          ERROR(
+              "The background does not provide an analytic solution. Select an "
+              "analytic solution as background if you want to use it to impose "
+              "boundary conditions.");
```
Comment (nilsdeppe) [inline src/Elliptic/BoundaryConditions/AnalyticSolution.hpp]: should this say `analytic boundary conditions`? Currently it reads like you cannot impose boundary conditions at all if the background doesn't have an analytic solution. I'm also confused why this is the first time `background` shows up. It's not clear what's special about the background here, vs the solution variables. My understanding was you impose boundary conditions on the solution variables. This seems like it's meant to be a user-facing error, if that is the case this needs to be a lot clearer and should maybe use `ERROR_NO_TRACE` instead

File: src/Elliptic/BoundaryConditions/AnalyticSolution.hpp [hunk 5]
```diff
@@ -0,0 +1,234 @@
...
+        }
+        return *optional_analytic_solutions;
+      } else {
+        return optional_analytic_solutions;
+      }
+    }();
+    const size_t slice_index =
+        index_to_slice_at(volume_mesh.extents(), direction);
+    const auto helper = [this, &analytic_solutions, &volume_mesh, &direction,
```
Comment (nilsdeppe) [inline src/Elliptic/BoundaryConditions/AnalyticSolution.hpp]: `helper` -> `...` whatever this actually does. Maybe `data_on_boundary` or something like that

File: src/Elliptic/BoundaryConditions/AnalyticSolution.hpp [hunk 6]
```diff
@@ -0,0 +1,234 @@
...
+  }
+
+  using argument_tags_linearized = tmpl::list<>;
+  using volume_tags_linearized = tmpl::list<>;
+
+  void apply_linearized(
+      const gsl::not_null<typename FieldTags::type*>... fields,
+      const gsl::not_null<typename FieldTags::type*>... n_dot_fluxes)
+      const noexcept {
+    const auto helper = [this](auto field_tag_v, const auto field,
```
Comment (nilsdeppe) [inline src/Elliptic/BoundaryConditions/AnalyticSolution.hpp]: again, rename

File: src/Elliptic/BoundaryConditions/AnalyticSolution.hpp [hunk 7]
```diff
@@ -0,0 +1,234 @@
...
+          ERROR("Unsupported boundary condition type: "
+                << get<elliptic::Tags::BoundaryConditionType<field_tag>>(
+                       boundary_condition_types()));
+      }
+    };
+    EXPAND_PACK_LEFT_TO_RIGHT(helper(FieldTags{}, fields, n_dot_fluxes));
+  }
+
+  // NOLINTNEXTLINE
+  void pup(PUP::er& p) noexcept override {
```
Comment (nilsdeppe) [inline src/Elliptic/BoundaryConditions/AnalyticSolution.hpp]: move out of line

File: tests/Unit/Elliptic/BoundaryConditions/Test_AnalyticSolution.cpp [hunk 8]
```diff
@@ -0,0 +1,145 @@
...
+              boundary_condition_base.get()) != nullptr);
+  const auto& boundary_condition =
+      dynamic_cast<const BoundaryConditionType&>(*boundary_condition_base);
+  test_serialization(boundary_condition);
+  test_copy_semantics(boundary_condition);
+  auto move_boundary_condition = boundary_condition;
+  test_move_semantics(std::move(move_boundary_condition), boundary_condition);
+
+  // Test applying the boundary conditions
+  const Mesh<1> mesh{3, Spectral::Basis::Legendre,
```
Comment (nilsdeppe) [inline tests/Unit/Elliptic/BoundaryConditions/Test_AnalyticSolution.cpp]: please add a test for at least 2d, ideally 3d too. 1d is very special/simple since there's only a point on the boundary.

**Review by nilsdeppe — CHANGES_REQUESTED** [review]

File: src/Elliptic/BoundaryConditions/AnalyticSolution.hpp (hunk 1 above)
Comment (nilsvu) [reply src/Elliptic/BoundaryConditions/AnalyticSolution.hpp:103]: Yea I need to have something to expand. Added a comment.

File: src/Elliptic/BoundaryConditions/AnalyticSolution.hpp (hunk 2 above)
Comment (nilsvu) [reply src/Elliptic/BoundaryConditions/AnalyticSolution.hpp]: Prepended a commit that switches the argument order

File: src/Elliptic/BoundaryConditions/AnalyticSolution.hpp (hunk 3 above)
Comment (nilsvu) [reply src/Elliptic/BoundaryConditions/AnalyticSolution.hpp]: Without the `decltype(auto)` the lambda assumes the return type is `auto` and returns by value. With the `decltype(auto)` the lambda returns by const-ref. This is assuming that I got this C++ rule right.

Comment (nilsdeppe) [reply src/Elliptic/BoundaryConditions/AnalyticSolution.hpp]: Okay, would `const auto&`as a return type work? That might be clearer

Comment (nilsvu) [reply src/Elliptic/BoundaryConditions/AnalyticSolution.hpp]: Yep that works. I think `decltype(auto)` is the right choice for generic code where the return value should be forwarded exactly.

**Comment by nilsvu** [comment]
@nilsdeppe @fmahebert Fixup posted and prepended a commit, please take a look. Cheers!

**Review by nilsdeppe — COMMENTED** [review]
LGTM, feel free to squash

**Review by fmahebert — COMMENTED** [review]
👍

**Comment by nilsvu** [comment]
Rebased and squashed. Thanks for reviewing @nilsdeppe and @fmahebert!

**Review by fmahebert — APPROVED** [review]

**Review by nilsdeppe — APPROVED** [review]
